GSN basics guide (#1847)
OpenZeppelin/openzeppelin-contracts: docs/modules/ROOT/pages/gsn.adoc

= The Gas Station Network

User onboarding is one of the hottest topics in Ethereum, and meta-transactions the de-facto solution to this important problem. Here you'll learn all about the Gas Station Network (GSN) and how to write contracts that don't require their users to hold Ether to pay for gas.

If you're already up to speed with the workings of the GSN, feel free to skip to [receiving-a-relayed-call], where we'll go over how to use the OpenZeppelin Contracts to easily write a GSNRecipient contract. Otherwise, strap in!

== Sending gas-less transactions

> All Ethereum transactions spend gas, and the sender of the transaction must have enough Ether to pay for these gas costs. Even though these costs are low for basic transactions (a couple cents), getting this Ether is no easy task: you'll probably need to go through Know Your Customer and Anti Money-Laundering processes (KYC & AML), which not only take time but often involve sending a selfie holding your passport over the Internet (!). Only the most hardcore users will put up with this hassle, and dApp adoption greatly suffers when Ether is required. We can do better.

Reviewer: Loved the intro! The ending of the first sentence looks odd though: it's hard to follow what _these gas costs_ refer to. Also, be careful with the change of subject: in the first sentence it's "the sender of the tx", in the second it's "you", and the third is "the most hardcore users".

> **Enter meta-transactions**. This is a fancy name for a simple idea: a third-party can send another user's transactions and pay themselves for the gas cost. That's it! There's some tricky technical details, but those can be safely ignored when interacting with the GSN. This way, instead of your users calling into your contract (called the _recipient_) directly, someone else (we'll call them a _relay_) will send their transaction and pay for the cost.

Reviewer: Are we using "relay" or "relayer"?

Reviewer: Also, let's clarify the reader that they don't need to run a relayer themselves: they are run by other parties in the network. This is one of the key differences with other meta-tx solutions.

Author: I worry that may confuse a majority of users not familiar with the details behind other solutions (for one, it would confuse me). Perhaps the distinction you mentioned is better suited to a high-level explanation of the GSN that compares it to other systems?

The following sentences immediately refer to the relayer's incentives, why they are paid, trusting them, etc., and generally speak of 'they', which I think is clear enough for this context.

But why would they do such a thing?

== Incentives

Relays are not running a charity: they're running a bussiness. The reason why they'll gladly pay for your users' gas costs is because they will in turn charge your contract, the recipient. That way relays get their money back, plus a bit extra as a _fee_ for their services.

> This may sound strange at first, but paying for user onboarding is a very common bussiness practice. Lots of money are spent on advertising, free trials, new user discounts, etc., all with the goal of user acquisition. And compared to those, the cost of a couple Ethereum transactions is actually very small.

Reviewer: If you want to throw in the technical term, you can refer to Customer Acquisition Cost.

Author: I'd rather not go all buzzwordy, though a link to e.g. wikipedia might be useful to give people some more context.

== One contract to coordinate them all

There are many meta-transaction implementations out there, but the GSN has a unique detail that makes it special. Sitting in the middle of the system, a smart contract is charged with keeping track of relays, handling relayed transactions, charging their recipients, and generally ensuring all parties stay honest. This contract is called RelayHub, and there is a _single_ instance of it in the whole network (you don't need to deploy your own!). Think of it as a piece of public infrastructure, for all Ethereum users to benefit from.

One of RelayHub's tasks is relay-coordination: all relays will advertise their services on this contract, and your users will query it to find the relay that best suits their purposes. This is out of scope for this guide however, and is not something you need to worry about when writing a recipient contract. If you want to learn more about sending transactions via relays, head to our [GSNProvider guide]().

== Receiving a relayed call

> The contract you will be inheriting from is GSNRecipient. It includes the IRelayRecipient interface, plus a number of useful utilities that handle the low-level details of the GSN, so you can focus on your dApp.

Reviewer: This section feels disconnected from the previous ones, and its beginning is a bit harsh. Why do I want to inherit from what? We should add a bit more context: tell the reader that we will now explain how to add GSN support in their app.

Author: I first wrote this bit and didn't review the transition when I added the intro, it did end up quite jarring. I'll add a couple sentences for context.

> Unlike regular contract function calls, each relayed call has an additional number of steps it must go through. Most of writing a recipient involves handling these, which are designed to provide enough flexibility to cater to all possible use cases. Basic recipients can however safely ignore most of them, while still being secure and sound.

Reviewer: ```suggestion
Unlike regular contract function calls, each relayed call has an additional number of steps it must go through. Most of writing a recipient involves handling these, which are designed to provide flexibility. Basic recipients can however safely ignore most of them, while still being secure and sound.
```

Reviewer: Also, what is a "relayed call"? This is the first time it's mentioned, and a reader new to this doesn't necessarily know about it.

The OpenZeppelin Contracts provide a number of [tried-and-tested approaches]() for you to use out of the box, but you should still have a basic idea of what's going on under the hood.

==== acceptRelayedCall

> First, RelayHub will ask your recipient contract if it wants to receive a relayed call. Recall that you will be charged for incurred gas costs by the relay, so you should only accept calls that you're willing to pay for!

Reviewer: Why would RelayHub ask my recipient contract? We are assuming here that the user knows the meta-tx flow in which the relayer calls into the hub, which in turn calls into the recipient. We may want to make that flow explicit before going into the details of each method.

There are multiple ways to make this work, including:
 * having a whitelist of trusted users
 * only accepting calls to an onboarding function
 * charging users in tokens (possibly issued by you)
 * delegating the acceptance logic off-chain

All relayed call requests can be rejected at no cost to the recipient.

=== pre and postRelayedCall

> You may have noticed one of the suggested acceptance flows involves charging a user in tokens. You may, for example, setup a system where users spend a special token to pay for relayed calls, and then either gift these to new users, or sell them off-chain. The GSN will let you easily implement these designs.

Reviewer: I'd move the example _after_ we have explained what pre and post are, or even remove this in favor of a link to the advanced use cases guide.

Author: The issue is that all pre and post do is allow for user charging, so we should definitely cover it beforehand.

I see two possible ways to tackle this:
 1. go over the 'user charging' schemes earlier in the guide (there's currently a short snippet in the 'incentives' section)
 2. make pre and post optional by having GSNRecipient include empty implementations, and only go over them in the advanced used cases guide

Reviewer: > The issue is that all pre and post do is allow for user charging, so we should definitely cover it beforehand.

True, but then I would start this paragraph with exactly that, instead of a direct reference to token payments. Otherwise, a reader may understand that it can only be used with an associated token.

I'd reorder this as follows:
1- Start with "After a relayed call is accepted...", since it ties very well with the previous section
2- Explain that these functions are usually used to charge your users, and that an example could be paying with a native token
3- Then go into the details of pre and post
4- Then some sample code

After a relayed call is accepted, RelayHub will give your contract two opportunities to charge your user for their call, perform some bookeeping, etc.: _before_ and _after_ the actual relayed call is made. These functions are aptly named preRelayedCall and postRelayedCall.

> preRelayedCall will inform you of the maximum cost the call may have, and is used to charge the user in advance. This is only required in scenarios where your users may run off with their tokens before you're able to charge them, and not all applications need this function.

Reviewer: We can make a parallel to how Ethereum handles gas: it first locks `gasPrice * gasLimit` for the tx, and then deducts the balance actually spent.

> postRelayedCall will give you an accurate estimate of the transaction cost, making it a natural place to charge users. It will also let you know if the relayed call reverted or not, which allows for more sophisticated interactions, like not charging users for reverted calls.

Reviewer: ```suggestion
postRelayedCall will give you an accurate estimate of the transaction cost, making it a natural place to charge users. It will also let you know if the relayed call reverted or not, which allows for more sophisticated interactions, like not charging users for reverted calls (but remember that your app will be charged by the relay nonetheless).
```

== msg.sender and msg.data

> There's only one extra care you should take when working with GSN recipient contracts: _you must never use msg.sender or msg.data directly_. Understanding why requires an in-depth analysis of the inner workings of the GSN and RelayHub, but you need not worry about these low-level details: GSNRecipient provides two functions, `_msgSender()` and `_msgData()`, which are drop-in replacements for `msg.sender` and `msg.data`. As long as you use these two functions instead, you're good to go!

Reviewer: > Understanding why requires an in-depth analysis of the inner workings of the GSN and RelayHub, but you need not worry about these low-level details

Let's replace this with a link to an explanation why this is the case. We don't need to go into details on how data and sender are actually serialized, just explain that the data & sender of a wrapped tx are not the same as those of the wrapper tx.

== Payment

> By now you may be wondering how exactly relays charge their recipients for gas costs and service fee. The answer is simple: the recipient must deposit funds on RelayHub in advance, and payment is automatically handled on each relayed call.

Reviewer: ```suggestion
By now you may be wondering how exactly relays charge their recipients for gas costs and service fee. The answer is simple: the recipient must deposit funds on the RelayHub in advance, and payment is automatically handled on each relayed call.
```

You can head to the [GSN dashboard](deltacamp.website) to check and top-up your contracts' balance, view previous charges, or you can do this programatically by calling IRelayHub.depositFor and IRelayHub.balanceOf.

> Recipients may withdraw their balance from the system at any point, but remember that they will not be able to receive any further relayed calls!

Reviewer: Again, be careful with the subject: in the previous paragraph the recipient was "you", but on this one it's "they".

Author: This comes from the fact that it is the recipient itself (the contract) that must call `withdraw`, there's no notion of an 'owner' of a recipient, and I didn't want to mislead the reader. 

Perhaps it'd be best to briefly mention this, i.e. "don't forget to add a way to call withdraw from your contract!".

Reviewer: > This comes from the fact that it is the recipient itself (the contract) that must call withdraw,

Damn, I wasn't taking this into consideration. This kinda forces you to have an owner of the contract who can call into withdraw.

Let's rephrase this sentence so it's clear that it is the contract who needs to withdraw, and add something (maybe an internal `withdraw` function?) in the GSNRecipientBase if we don't have one.

= The Gas Station Network

> User onboarding is one of the hottest topics in Ethereum, and meta-transactions the de-facto solution to this important problem. Here you'll learn all about the Gas Station Network (GSN) and how to write contracts that don't require their users to hold Ether to pay for gas.

Reviewer 2: Should "and meta-transactions the de-facto solution"
be "and meta-transactions is the de-facto solution"?

If you're already up to speed with the workings of the GSN, feel free to skip to <<Receiving a relayed call>>, where we'll go over how to use the OpenZeppelin Contracts to easily write a GSNRecipient contract. Otherwise, strap in!

== Sending gas-less transactions

> All Ethereum transactions use gas, and the sender of each transaction must have enough Ether to pay for the gas spent. Even though these gas costs are low for basic transactions (a couple cents), getting this Ether is no easy task: dApp users often need to go through Know Your Customer and Anti Money-Laundering processes (KYC & AML), which not only take time but often involve sending a selfie holding their passport over the Internet (!). Only the most hardcore users will put up with this hassle, and dApp adoption greatly suffers when Ether is required. We can do better.

Reviewer 2: "a couple cents" could be "a couple of cents"

"which not only take time but often involve sending a selfie holding their passport over the Internet (!)" could be
"which not only takes time but often involves sending a selfie holding their passport over the Internet(!)"

**Enter meta-transactions**. This is a fancy name for a simple idea: a third-party can send another user's transactions and pay themselves for the gas cost. That's it! There's some tricky technical details, but those can be safely ignored when interacting with the GSN. This way, instead of your users calling into your contract (called the _recipient_) directly, someone else (we'll call them a _relayer_) will send their transaction and pay for the cost.

But why would they do such a thing?

== Incentives

> Relayers are not running a charity: they're running a bussiness. The reason why they'll gladly pay for your users' gas costs is because they will in turn charge your contract, the recipient. That way relayers get their money back, plus a bit extra as a _fee_ for their services.

Reviewer 2: typo x 2: "bussiness" should be "business"

"Lots of money are" could be "Lots of money is"

"And compared to those, the cost of a couple Ethereum transactions is actually very small" could be
"Compared to those, the cost of a couple of Ethereum transactions is actually very small"

This may sound strange at first, but paying for user onboarding is a very common bussiness practice. Lots of money are spent on advertising, free trials, new user discounts, etc., all with the https://en.wikipedia.org/wiki/Customer_acquisition_cost[goal of user acquisition]. And compared to those, the cost of a couple Ethereum transactions is actually very small.

Additionally, you can leverage the GSN in scenarios where your users pay you off-chain in advance (e.g. via credit card), with each GSN-call deducting from their balance on your system. The possibilities are endless!

=== Should I trust these relayers?

> You don't need to! The GSN is set up in such a way where it's in the relayers' best interest to serve your requests, and there are measures in place to penalize them if they misbeheave. All of this happens automatically, so you can safely start using their services worry-free.

Reviewer 2: typo "misbeheave" should be "misbehave"

One of RelayHub's jobs is to act as a, well, _hub_ for all relayers: they will advertise their services on this contract, and your users will query it to find the relayer that best suits their purposes. This is out of scope for this guide however, and is not something you need to worry about when writing a recipient contract. If you want to learn more about sending transactions via relayers, head to our [GSNProvider guide]().

The other key task RelayHub carries out is the actual _relaying of transactions_, the sole purpose behind this whole system. Instead of calling a function in your contract directly, your users will request a relayer to do it for them, who will then execute RelayHub's relayCall function. RelayHub will verify that the transaction is legitimate (protecting both users and recipients from dishonest relayers), and then call into your contract as originally requested by your user. This requires your recipient trusting RelayHub to do the right thing, but since it is a smart contract, this is as simple as reading its source code!

== Receiving a relayed call

> We've mentioned how it is RelayHub, not your user, that actually ends up calling a function in your contract. We will refer to this as the _relayed call_. OpenZeppelin Contracts includes a number of utilities to make receiving relayed calls as easy as developing a regular Solidiy contract, without needing to worry about the low level details.

Reviewer 2: typo "Solidiy" should be "Solidity"

The first step to writing a recipient is to inherit from our GSNRecipient contract. If you're also inheriting from other OpenZeppelin contracts, such as ERC20 or Crowdsale, this will work just fine: by adding GSNRecipient all of your token or crowdsale functions will also be GSN-callable.

```solidity
import "@openzeppelin/contracts/GSN/GSNRecipient.sol";

contract MyContract is GSNRecipient, ... {

}
```

=== msg.sender and msg.data

> There's only one extra care you should take when working with GSN recipient contracts: _you must never use `msg.sender` or `msg.data` directly_. On relayed calls, `msg.sender` will be `RelayHub` instead of your user! This doesn't mean however you won't be able to retrieve your users' addresses: `GSNRecipient` provides two functions, `_msgSender()` and `_msgData()`, which are drop-in replacements for `msg.sender` and `msg.data` while taking care of the low-level details. As long as you use these two functions instead of the original getters, you're good to go!

Reviewer 2: "There's only one extra care you should take when working with GSN recipient contracts" could be
"There's only one extra detail you need to take care of when working with GSN recipient contracts"

"As long as you use these two functions instead of the original getters"
Are `msg.sender` and `msg.data` getters?

Author: Technically I guess they could be considered properties of a magical global `msg` struct. I dislike the 'getters' term though, do you have any suggestions?

Reviewer 2: I don't have a better term than getters, I had a quick look through the solidity documentation but didn't see anything useful.

Perhaps just avoid using a term and specify them directly.

```suggestion
As long as you use these two replacement functions instead of `msg.sender` and `msg.data`, you're good to go!
```

> WARNING: Third-party contracts you inherit from may not use these replacement functions, making them unsafe to use when mixed with `GSNRecipient`. If in doubt, head on over to our https://forum.openzeppelin.com/c/support[support forum].

Author: The 'WARNING' bit will render to this: https://imgur.com/a/Zw7LCyE

=== Accepting and charging

> Unlike regular contract function calls, each relayed call has an additional number of steps it must go through, which are functions of the `IRelayRecipient` interface `RelayHub` will call on your contract. `GSNRecipient` includes this interface but no implementation: most of writing a recipient involves handling these function calls. They are designed to provide flexibility, but basic recipients can safely ignore most of them while still being secure and sound.

Reviewer 2: "secure and sound" could be "safe and sound" or "safe and secure"

The OpenZeppelin Contracts provide a number of [tried-and-tested approaches]() for you to use out of the box, but you should still have a basic idea of what's going on under the hood.

> ==== acceptRelayedCall

Reviewer: WDYT about adding the code for this function, with a sample `return true` implementation? 
I think it may make this easier to understand.

First, RelayHub will ask your recipient contract if it wants to receive a relayed call. Recall that you will be charged for incurred gas costs by the relayer, so you should only accept calls that you're willing to pay for!

There are multiple ways to make this work, including:

>  * having a whitelist of trusted users

Reviewer 2: Bullet points need to be formatted.
"* having a whitelist of trusted users * only accepting calls to an onboarding function * charging users in tokens (possibly issued by you) * delegating the acceptance logic off-chain"

Author: Not sure what you mean by formatting, could you expand?

Reviewer 2: The bullet points didn't format as bullet points in the preview of the document.  

The preview document displays as:
"* having a whitelist of trusted users * only accepting calls to an onboarding function * charging users in tokens (possibly issued by you) * delegating the acceptance logic off-chain"

Rather than:
* having a whitelist of trusted users 
* only accepting calls to an onboarding function 
* charging users in tokens (possibly issued by you) 
* delegating the acceptance logic off-chain

Author: Oh nice, thanks! I'll look into how lists are made in asciidoc. Where did you find the deploy preview? I though those were only generated for PRs to master.

Reviewer 2: In Files Changed in the PR you can view the file and that shows the formatted document

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/ecb569b23804ad3b56e700e59a4dff8e94a4a0a8/docs/modules/ROOT/pages/gsn.adoc

Author: Oh I see, I didn't know github rendered asciidoc. Thanks!

After a relayed call is accepted, RelayHub will give your contract two opportunities to charge your user for their call, perform some bookeeping, etc.: _before_ and _after_ the actual relayed call is made. These functions are aptly named preRelayedCall and postRelayedCall.

preRelayedCall will inform you of the maximum cost the call may have, and is used to charge the user in advance. This is only required in scenarios where your users may run off with their tokens before you're able to charge them, and not all applications need this function.

postRelayedCall will give you an accurate estimate of the transaction cost, making it a natural place to charge users. It will also let you know if the relayed call reverted or not, which allows for more sophisticated interactions, like not charging users for reverted calls (but remember that your recipient contract will be charged by the relayer nonetheless!).

== Payment

> By now you may be wondering how exactly relayers charge their recipients for gas costs and service fee. The answer is simple: each recipient must have funds deposited on RelayHub in advance, and payment is automatically handled on each relayed call.

Reviewer 2: "service fee" could be "service fees" or "the service fee"
